### ISSUE TITLE
c_mask can be negative?

### ISSUE THREAD

**Issue by xiong224** [shleecs/Copy-and-Paste-Networks-for-Deep-Video-Inpainting#8]
Hello, thanks for your excellent work. I am a student and begin to do the video inpainting. When I run the code, I find c_mask may be negative sometimes. However, according to the function described in the paper, it should not be negative. Could you tell me where i am wrong? Thanks very much!

**Comment by shleecs** [comment]
Hello,
Thank you for your interest in our paper.
I also agree with you, the c_mask can not be negative, and the code also cannot generate negative value.

Please check our code in the CPNet_model.py.
1. vmaps
rvmap_t = (F.upsample(rvmaps[:,:,r], size=(H, W), mode='bilinear', align_corners=False)>0.5).float()
vmap_.append(rvmap_t)
vmaps = torch.stack(vmap_, dim=2) (vmaps > 0)

2. c_match
def masked_softmax(self, vec, mask, dim):
        masked_vec = vec * mask.float()
        max_vec = torch.max(masked_vec, dim=dim, keepdim=True)[0]
        exps = torch.exp(masked_vec-max_vec) # exps: exponential value is always greater than 0.
        masked_exps = exps * mask.float() # mask: vmaps, therefore the mask value is greater than 0.
        masked_sums = masked_exps.sum(dim, keepdim=True) 
        zeros = (masked_sums <1e-4)        
        masked_sums += zeros.float()
        return masked_exps/masked_sums # Thus, the return value is greater than zero.
c_match = self.masked_softmax(gss, vmaps, dim=2)

3. c_mask
c_mask = (c_match * vmaps) # The summation of vmaps in dimension=2 is smaller than 1.
c_mask = torch.sum(c_mask,2) 
c_mask = 1. - (torch.mean(c_mask, 1, keepdim=True)) # Therefore, final c_mask is also smaller than 1.

Still, if you have concern our code as the same reason, please add the c_mask = (c_mask > 0.).float() in the CPNet_model.py 218L.
Thank you.
